File: 백준/Gold/2636. 치즈/치즈.py
import sys
from collections import deque
input = sys.stdin.readline

def bfs():
    que = deque()
    que.append((0, 0))
    visited[0][0] = True
    cnt = 0

    while que:
        x, y = que.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < r and 0 <= ny < c and not visited[nx][ny]:
                visited[nx][ny] = True
                if graph[nx][ny] == 0:
                    que.append((nx, ny))
                elif graph[nx][ny] == 1:
                    graph[nx][ny] = 0
                    cnt += 1

    return cnt


r, c = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(r)]
dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]
ans = []
time = 0

while True:
    visited = [[False] * c for _ in range(r)]
    cnt = bfs()

    if cnt == 0:
        break

    time += 1
    ans.append(cnt)

print(time)
print(ans[-1])

File: 백준/Gold/2636. 치즈/치즈.java
import java.io.*;
import java.util.*;

public class Main {
	static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder output = new StringBuilder();
	static StringTokenizer tokens;
	static int N, M, time, ans, cheese;
	static int[][] map, deltas;
	static boolean[][] visited;

	public static void main(String[] args) throws IOException {
		tokens = new StringTokenizer(input.readLine());
		N = Integer.parseInt(tokens.nextToken());
		M = Integer.parseInt(tokens.nextToken());
		map = new int[N][M];
		deltas = new int[][] {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
		
		for(int r = 0; r < N; r++) {
			tokens = new StringTokenizer(input.readLine());
			for(int c = 0; c < M; c++) {
				map[r][c] = Integer.parseInt(tokens.nextToken());
				if(map[r][c] == 1) cheese++;
			}
		}
		
		while(cheese != 0) {
			visited = new boolean[N][M];
			time++;
			ans = cheese;
			bfs();
		}
		
		System.out.println(time);
		System.out.println(ans);
	}
	
	static void bfs() {
		Deque<Pos> que = new ArrayDeque<>();
		que.offer(new Pos(0, 0));
		visited[0][0] = true;
		
		while(!que.isEmpty()) {
			Pos now = que.poll();
			int r = now.r;
			int c = now.c;
			
			for(int d = 0; d < deltas.length; d++) {
				int nr = r + deltas[d][0];
				int nc = c + deltas[d][1];
				
				if(nr < 0 || nr >= N || nc < 0 || nc >= M || visited[nr][nc]) {
					continue;
				}
				
				visited[nr][nc] = true;
				if(map[nr][nc] == 0) {
					que.offer(new Pos(nr, nc));
				} else {
					cheese--;
					map[nr][nc] = 0;
				}
			}
		}
	}
	
	static class Pos {
		int r;
		int c;
		
		public Pos(int r, int c) {
			super();
			this.r = r;
			this.c = c;
		}
	}
}